modules/oe_translation_corporate_workflow/tests/src/FunctionalJavascript/CorporateWorkflowTranslationDropTest.php

```diff
@@ -0,0 +1,314 @@
+    $node = $node_storage->create([
+      'type' => 'page',
+      'title' => 'My node',
+      'moderation_state' => 'draft',
+    ]);
+    $node->save();
+
+    $this->drupalGet($node->toUrl());
+
+    // Publish the content. We use the shortcuts for that.
+    $this->getSession()->getPage()->selectFieldOption('Change to', 'published');
+    $this->getSession()->getPage()->pressButton('Apply');
+    $this->waitForBatchExecution();
+    $this->assertSession()->waitForText('The moderation state has been updated.');
+
+    $this->assertRevisions('node', 5);
+    $this->assertRevisions('content_moderation_state', 5);
+
+    // Translate the content.
+    $request = $this->createLocalTranslationRequest($node, 'fr');
+    $this->drupalGet($request->toUrl('local-translation'));
+    $this->assertSession()->elementContains('css', '#edit-title0value-translation', 'My node');
+    $values = [
+      'Translation' => 'My node FR',
+    ];
+    $this->submitForm($values, t('Save and synchronise'));
+
+    // We still only 5 revisions of each and only the published one has a
+    // translation.
+    $this->assertRevisions('node', 5, [5]);
+    $this->assertRevisions('content_moderation_state', 5, [5]);
+
+    // Create a new draft.
+    $node_storage->resetCache();
+    $node = $node_storage->load($node->id());
+    $node->set('title', 'My node 2');
+    $node->set('moderation_state', 'draft');
+    $node->save();
+
+    // We now have 6 revisions, and the last two have translations as the
+    // translation gets carried over from the published revision to the draft.
+    $this->assertRevisions('node', 6, [5, 6]);
+    $this->assertRevisions('content_moderation_state', 6, [5, 6]);
+
+    // Go the node and publish it using the shortcuts.
+    $this->drupalGet($node->toUrl('latest-version'));
+    $this->getSession()->getPage()->selectFieldOption('Change to', 'published');
+    $this->getSession()->getPage()->checkField('Do not carry over the translations');
+    $this->getSession()->getPage()->pressButton('Apply');
+    $this->waitForBatchExecution();
+    $this->assertSession()->waitForText('The moderation state has been updated.');
+    $this->assertRevisions('node', 10, [5, 6, 7, 8]);
+    // Due to content moderation also handling the deletion of translations,
+    // we lose the translation of the "request_validation" revision.
+    // @see oe_translation_corporate_workflow_content_moderation_state_presave().
+    $this->assertRevisions('content_moderation_state', 10, [5, 6, 7]);
+  }
+
+  /**
+   * Asserts the entity revisions are as we expect them.
+   *
+   * It assumes the node to have the ID of 1.
+   *
+   * @param string $entity_type
+   *   The entity type ID.
+   * @param int $count
+   *   The number of revisions we expect.
+   * @param array $with_translations
+   *   The revision IDs we expect to have a FR translation.
+   */
+  protected function assertRevisions(string $entity_type, int $count, array $with_translations = []) {
```

Comment:
Can this method be moved in the `TranslationsTestTrait` so it can be used in `CorporateWorkflowTranslationTest` too?

Reply:
It's ok, let's leave it.